### PR TITLE
STORM-3392 Topology page should show components even if workers haven…

### PR DESCRIPTION
**Pull request by dandsager1**
…'t started

### CODE REVIEW

**Comment by dandsager1** [comment]
Topology page sample
Screen Shot 2019-05-16 at 10 26 18 AM

**Comment by srdo** [comment]
Looks good, but could you elaborate on why it is useful to show these placeholders?

**Comment by dandsager1** [comment]
This is useful when there are no metrics, which happens when starting up a topology.

**Comment by srdo** [comment]
I meant why is it useful? I get that it has an effect when the topology starts, but why is it useful to you that the placeholders are shown, rather than the UI being empty?

**Comment by d2r** [comment]
@srdo I think in a multi-tenant environment, where some of us have to try and diagnose another user's topology that is not working well, it is would be nice if the UI at least displays the topology's components regardless of whether they have yet sent any metrics. At least that is a case when I would find it useful.

In fact it would be nice to enhance the Topology Visualization in the same way, but that probably is a more involved change and might best be done separately.

**Comment by srdo** [comment]
@d2r Sounds good, thanks for explaining. +1.

File: storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java [hunk 1]
```diff
@@ -4129,6 +4136,83 @@ public TopologyPageInfo getTopologyPageInfo(String topoId, String window, boolea
...
+     * Add placeholder AggStats allowing topology page to show components before AggStats are populated.
+     *
+     * @param topoPageInfo  topology page info holding bolt AggStats
+     * @param topology      storm topology used to get bolt names
+     */
+    private void maybeAddPlaceholderBoltAggStats(TopologyPageInfo topoPageInfo, StormTopology topology) {
+        if (topoPageInfo.get_id_to_bolt_agg_stats().isEmpty()) {
+            Map<String, Bolt> bolts = topology.get_bolts();
+            for (Entry<String, Bolt> entry : bolts.entrySet()) {
+                if (Utils.isSystemId(entry.getKey())) {
```
Comment (Ethanlm) [inline storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java]: Sorry I am too late to review. But why do we want to skip system bolts?

Comment (dandsager1) [reply storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java]: The idea was to show active topology components. It is easy to filter based on the includeSys param passed into getTopologyPageInfo

**Pull request merged**